curated_transformers/models/loading_util.py

```diff
@@ -0,0 +1,105 @@
+
+    size_mismatch_msg = (
+        "Expected the size of the replacement for `{}` to be {}, but got {}"
+    )
+
+    if is_parameter:
+        old_param = module._parameters[tensor_name]
+        assert old_param is not None
+        if old_param.shape != tensor.shape:
+            raise ValueError(
+                size_mismatch_msg.format(tensor_name, old_param.shape, tensor.shape)
+            )
+        new_param = Parameter(tensor, requires_grad=old_param.requires_grad)
+        module._parameters[tensor_name] = new_param
+    else:
+        old_buffer = module._buffers[tensor_name]
+        assert old_buffer is not None
+        if old_buffer.shape != tensor.shape:
+            raise ValueError(
+                size_mismatch_msg.format(tensor_name, old_buffer.shape, tensor.shape)
+            )
+        module._buffers[tensor_name] = tensor
+
+
+def emplace_module_state_dict(
+    module: Module,
+    state_dict: Mapping[str, torch.Tensor],
+    *,
+    device: Optional[torch.device] = None,
+):
+    """Recursively replace a module's parameters and buffers with the tensors in the given
+    `state_dict` without performing any unnecessary copies.
+
+    :param module: Top-level module whose child tensors needs replacing.
+    :param state_dict: State dictionary containing the replacement tensors
+        and their names.
+    :param device: Device to which the replacement tensors are moved.
+    """
+
+    def emplace(module: Module, state_dict: Mapping[str, torch.Tensor], prefix: str):
+        persistent_buffers = {
+            k: v
+            for k, v in module._buffers.items()
+            if k not in module._non_persistent_buffers_set
+        }
+        local_name_params = itertools.chain(
+            module._parameters.items(), persistent_buffers.items()
+        )
+
+        for name, param in local_name_params:
+            key = prefix + name
+            if param is None:
+                if key in state_dict:
+                    raise ValueError(
+                        f"Key `{name}` found in state dict but no data in module `{prefix}`"
+                    )
+                else:
+                    continue
+            if key not in state_dict:
+                raise ValueError(f"Key `{name}` not found in module `{prefix}`")
+            replacement = state_dict[key]
+            _emplace_module_tensor(module, name, replacement, device=device)
+
+    def traverse(module: Module, state_dict: Mapping[str, torch.Tensor], prefix: str):
+        emplace(module, state_dict, prefix)
+
+        for name, child in module._modules.items():
+            if child is not None:
+                child_prefix = prefix + name + "."
+                child_state_dict = {
+                    k: v for k, v in state_dict.items() if k.startswith(child_prefix)
+                }
+                traverse(child, child_state_dict, child_prefix)
+
+    traverse(module, state_dict, "")
+    del traverse
+    del emplace
```

Comment:
Why are these needed? I could understand the deletion of `emplace` immediately after the `emplace` call (to avoid keeping it in memory while traversing, but doesn't do the interpreter the necessary analysis?)

Reply:
I was just following [what `Module.load_state_dict` does](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.load_state_dict), to be honest. I assumed it was to make sure that the `state_dict` wouldn't stick around, but it wouldn't make a difference in our case.

Reply:
Rewrote the function to use a queue instead (and added some extra sanity checks).

Reply:
Much nicer!